botocore/http_session.py

```
@@ -0,0 +1,216 @@
+                # If an http request is being proxied use the full url
+                request_target = request.url
+
+            conn = manager.connection_from_url(request.url)
+            self._setup_ssl_cert(conn, request.url, self._verify)
+            urllib_response = conn.urlopen(
+                method=request.method,
+                url=request_target,
+                body=request.body,
+                headers=request.headers,
+                retries=False,
+                assert_same_host=False,
+                preload_content=False,
+                decode_content=False,
+            )
+
+            http_response = botocore.awsrequest.AWSResponse(
+                request.url,
+                urllib_response.status,
+                urllib_response.headers,
+                urllib_response,
+            )
+
+            if not request.stream_output:
+                # Cause the raw stream to be exhausted immediately. We do it
+                # this way instead of using preload_content because
+                # preload_content will never buffer chunked responses
+                http_response.content
+
+            return http_response
+        except URLLib3SSLError as e:
+            raise SSLError(endpoint_url=request.url, error=e)
+        except (NewConnectionError, socket.gaierror) as e:
+            raise EndpointConnectionError(endpoint_url=request.url, error=e)
+        except ProxyError as e:
+            raise ProxyConnectionError(proxy_url=proxy_url, error=e)
+        except URLLib3ConnectTimeoutError as e:
+            raise ConnectTimeoutError(endpoint_url=request.url, error=e)
+        except URLLib3ReadTimeoutError as e:
+            raise ReadTimeoutError(endpoint_url=request.url, error=e)
+        except ProtocolError as e:
+            raise ConnectionClosedError(
+                error=e,
+                request=request,
+                endpoint_url=request.url
+            )
+        except Exception as e:
```

Are we sure we want to catch `Exception` here?  Is is possible to restrict this to a list of exceptions we know about?  This will consider even trivial bugs/typos as an `HTTPClientError`.

---

I think so, it ensures we never directly leak any exceptions from our dependency or a mistake in our implementation. What's your concern with an exception from this layer being wrapped as an `HTTPClientError`?

---

Primarily that actual bugs that raise things like `TypeError`/`ValueError` etc will be misrepresented as a `HTTPClientError`.  Ideally we'd want these things to fail fast and propagate all the way up the stack vs. possibly being retried and surfaced as a (somewhat confusing) `HTTPClientError`.

Note a huge deal, but I'd prefer to avoid it if possible.